fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -1100,23 +1139,50 @@ private void computeFragmentHosts() throws Exception {
                 if (hasShuffleHashBucketJoinInFragment) {
                     FragmentExecParams execParams = getMaxParallelismScanFragmentExecParams();
                     if (execParams != null) {
+                        maxParallelism = execParams.instanceExecParams.size();
+                        // when dop adaptation enabled, numInstances * pipelineDop is equivalent to numInstances in
+                        // non-pipeline engine and pipeline engine(dop adaptation disabled).
+                        if (dopAdaptionEnabled) {
+                            maxParallelism *= execParams.fragment.getPipelineDop();
+                        }
                         maxParallelismFragmentExecParams = execParams;
                     }
                 }
 
+                // hostSet contains target backends to whom fragment instances of the current PlanFragment will be
+                // delivered. when pipeline parallelization is adopted, the number of instances should be the size
+                // of hostSet, that it to say, each backend has exactly one fragment.
+                Set<TNetworkAddress> hostSet = Sets.newHashSet();
+                for (FInstanceExecParam execParams : maxParallelismFragmentExecParams.instanceExecParams) {
+                    hostSet.add(execParams.host);
+                }
+
+                if (dopAdaptionEnabled) {
+                    Preconditions.checkArgument(leftMostNode instanceof ExchangeNode);
+                    DataSink sink = getDataStreamSink(maxParallelismFragmentExecParams.fragment, fragment);
+                    Preconditions.checkArgument(sink != null);
+                    // If the maximum parallel child fragment send data to the current PlanFragment via UNPARTITIONED
+                    // DataStreamSink, then fragment instance parallelization is leveraged, otherwise, pipeline parallelization
+                    // is adopted.
+                    if (!sink.getOutputPartition().isPartitioned()) {
+                        // fragment instance parallelization (numInstances=N, pipelineDop=1)
+                        maxParallelism = Math.min(hostSet.size() * degreeOfParallelism, maxParallelism);
+                        fragment.setPipelineDop(1);
+                    } else {
+                        // pipeline parallelization (numInstances=|hostSet|, pipelineDop=degreeOfParallelism)
+                        maxParallelism = hostSet.size();
+                        fragment.setPipelineDop(degreeOfParallelism);
+                    }
```

Comment:
use `Fragment.getOutputPartition` is enough, sink is same with fragment

Reply:
ok

Reply:
if the maximum parallel child PlanFragment is type of MultiCastPlanFragment, Fragment.getOutputPartition is illegal.
```
// com/starrocks/planner/MultiCastPlanFragment.java#88
 public void setOutputPartition(DataPartition outputPartition) {
        Preconditions.checkState(false);
 }

```